File: chat-service/src/chat/chat.controller.ts
import {
  Body,
  Controller,
  Get,
  Param,
  Patch,
  Post,
  Query,
  UseGuards,
} from '@nestjs/common';
import { ChatService } from './chat.service';
import { InternalAuthGuard } from './internal-auth.guard';
import { CurrentUser } from './current-user.decorator';
import { CreateConversationDto } from './dto/create-conversation.dto';
import { SendMessageDto } from './dto/send-message.dto';
import { ListMessagesQueryDto } from './dto/list-messages-query.dto';

@UseGuards(InternalAuthGuard)
@Controller('chat')
export class ChatController {
  constructor(private readonly chatService: ChatService) {}

  @Get('conversations')
  listConversations(@CurrentUser() user: CurrentUser) {
    return this.chatService.listConversations(user);
  }

  @Post('conversations')
  createConversation(
    @Body() dto: CreateConversationDto,
    @CurrentUser() user: CurrentUser,
  ) {
    return this.chatService.createConversation(dto, user);
  }

  @Get('conversations/:conversationId/messages')
  listMessages(
    @Param('conversationId') conversationId: string,
    @Query() query: ListMessagesQueryDto,
    @CurrentUser() user: CurrentUser,
  ) {
    return this.chatService.listMessages(
      conversationId,
      user,
      query.limit,
      query.before,
    );
  }

  @Post('conversations/:conversationId/messages')
  sendMessage(
    @Param('conversationId') conversationId: string,
    @Body() dto: SendMessageDto,
    @CurrentUser() user: CurrentUser,
  ) {
    return this.chatService.sendMessage(conversationId, dto, user);
  }

  @Patch('conversations/:conversationId/read')
  markAsRead(
    @Param('conversationId') conversationId: string,
    @CurrentUser() user: CurrentUser,
  ) {
    return this.chatService.markAsRead(conversationId, user);
  }
}

File: chat-service-csharp/ChatService/Controllers/ChatController.cs
using ChatService.DTOs;
using ChatService.Security;
using ChatService.Services;
using Microsoft.AspNetCore.Mvc;

namespace ChatService.Controllers;

[ApiController]
[Route("chat")]
public class ChatController : ControllerBase
{
    private readonly IChatMessagingService _chatService;
    private readonly ICurrentUserService _currentUserService;

    public ChatController(
        IChatMessagingService chatService,
        ICurrentUserService currentUserService)
    {
        _chatService = chatService;
        _currentUserService = currentUserService;
    }

    [HttpGet("conversations")]
    public async Task<IActionResult> ListConversations()
    {
        var user = _currentUserService.Get();
        return Ok(await _chatService.ListConversationsAsync(user));
    }

    [HttpPost("conversations")]
    public async Task<IActionResult> CreateConversation([FromBody] CreateConversationDto dto)
    {
        var user = _currentUserService.Get();
        return Ok(await _chatService.CreateConversationAsync(dto, user));
    }

    [HttpGet("conversations/{conversationId}/messages")]
    public async Task<IActionResult> ListMessages(
        [FromRoute] string conversationId,
        [FromQuery] ListMessagesQuery query)
    {
        var user = _currentUserService.Get();
        return Ok(await _chatService.ListMessagesAsync(conversationId, user, query.Limit, query.Before));
    }

    [HttpPost("conversations/{conversationId}/messages")]
    public async Task<IActionResult> SendMessage(
        [FromRoute] string conversationId,
        [FromBody] SendMessageDto dto)
    {
        var user = _currentUserService.Get();
        return Ok(await _chatService.SendMessageAsync(conversationId, dto, user));
    }

    [HttpPatch("conversations/{conversationId}/read")]
    public async Task<IActionResult> MarkAsRead([FromRoute] string conversationId)
    {
        var user = _currentUserService.Get();
        return Ok(await _chatService.MarkAsReadAsync(conversationId, user));
    }
}
